File: Python/CircusBuffer_a.py


class CircusBuffer:

    def __init__(self, size):
        self.data = [0 for i in range(size)]
        self.size = size
        self.addIteration = 0
        self.getIteration = 0
        self.occupiedSpace = 0

    def GetFreeSpace(self):
        return self.size - self.occupiedSpace

    def GetOccupiedSpace(self):
        return self.occupiedSpace

    def Add(self, val):
        if self.occupiedSpace >= self.size:
            print("Buffer is full! Data was overwrited.")
            self.data[self.addIteration] = val
            self.addIteration += 1
            self.addIteration %= self.size
            self.getIteration += 1
        else:
            self.data[self.addIteration] = val
            self.addIteration += 1
            self.addIteration %= self.size
            self.occupiedSpace += 1

    def Get(self):
        if self.occupiedSpace == 0:
            print("Buffer is empty")
            return
        res = self.data[self.getIteration]
        self.getIteration += 1
        self.getIteration %= self.size
        self.occupiedSpace -= 1
        return res

size = 5
count = 7
buffer = CircusBuffer(size)

for i in range (0, count):
    buffer.Add(i);
    print("Добавили элемент: ", i ,"; Свободно памяти: ", buffer.GetFreeSpace())

for i in range (0, size):
    print("Достали элемент: ", buffer.Get() ,"; Свободно памяти: ", buffer.GetFreeSpace())

File: Cpp/CircusBuffer_a.cpp
#include <iostream>

using namespace std;

class CircularBuffer {
private:
    int* _data;
    int  _size;
    int _addIteration;
    int _getIteration;
    int  _occupiedSpace;

public:
    CircularBuffer(int size) {
        _occupiedSpace = 0;
        _getIteration = 0;
        _addIteration = 0;
        _size = size;
        _data = new int[_size];
    };

    ~CircularBuffer() {
        delete[] _data;
    };

    int GetFreeSpace() {
        return _size - _occupiedSpace;
    };

    int GetOccupiedSpace() {
        return _occupiedSpace;
    };

    void Add(int val) {
        if (_occupiedSpace >= _size) {
            cout << "Buffer is full! Data was overwrited." << endl;
            _data[_addIteration++] = val;
            _addIteration %= _size;
            _getIteration++;
        }
        else {
            _data[_addIteration++] = val;
            _addIteration %= _size;
            _occupiedSpace++;
        }
    };

    int Get() {
        if (_occupiedSpace == 0)
            throw "Буффер пуст";

        int res = _data[_getIteration++];
        _getIteration %= _size;
        --_occupiedSpace;
        return res;
    };
};

int main(int argc, char** argv)
{
    setlocale(0, "");
    int count = 7;
    CircularBuffer c(5);

    for (int i = 1; i <= count; ++i) {
        c.Add(i);
        cout << "Добавили элемент: " << i << endl;
        cout << "Всего элементов: " << c.GetOccupiedSpace() << endl;
    }

    count = c.GetOccupiedSpace();
    for (int i = 0; i < count; ++i) {
        cout << "Достали элемент: " << c.Get() << endl;
        cout << "Всего элементов: " << c.GetOccupiedSpace() << endl;
    }


    return 0;
}

